src/v/v8_engine/script_dispatcher.h

```diff
@@ -0,0 +1,146 @@
+
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <absl/container/node_hash_set.h>
+#include <v8_engine/script.h>
+
+namespace v8_engine {
+
+struct script_compile_params {
+    script_compile_params(ss::sstring function_name, iobuf code)
+      : function_name(std::move(function_name))
+      , code(std::move(code)) {}
+
+    ss::sstring function_name;
+    iobuf code;
+    size_t max_heap_size_bytes{10000};
+    size_t run_timeout_ms{200};
+};
+
+template<typename Executor>
+class script_dispatcher {
+public:
+    explicit script_dispatcher(Executor& executor)
+      : _executor(executor) {}
+
+    // Insert only put compilation params to map.
+    bool insert(
+      const model::topic_namespace& topic,
+      ss::sstring function_name,
+      iobuf code) {
+        return _compilation_params
+          .emplace(
+            topic,
+            script_compile_params(std::move(function_name), std::move(code)))
+          .second;
+    }
+
+    bool remove(const model::topic_namespace& topic) {
+        bool contains_script_flag = _compilation_params.erase(topic) == 1;
+        if (!contains_script_flag) {
+            return false;
+        }
+
+        auto runtime = _runtimes.find(topic);
+        if (runtime == _runtimes.end()) {
+            return true;
+        }
+
+        if (runtime->second.get()->function.get() == nullptr) {
+            runtime->second->is_deleted = true;
+            runtime->second->cv.broadcast();
+        }
+
+        _runtimes.erase(runtime);
+
+        return true;
+    }
+
+    ss::future<ss::lw_shared_ptr<script>> get(model::topic_namespace topic) {
+        auto params_it = _compilation_params.find(topic);
+        if (params_it == _compilation_params.end()) {
+            co_return ss::lw_shared_ptr<script>(nullptr);
+        }
+
+        auto runtime_it = _runtimes.find(topic);
+        if (runtime_it == _runtimes.end()) {
+            auto value = ss::make_lw_shared(runtimes_value());
+            _runtimes.emplace(value);
+
+            auto runtime = ss::make_lw_shared(script(
+              params_it->second.max_heap_size_bytes,
+              params_it->second.run_timeout_ms));
+
+            try {
+                co_await runtime->init(
+                  params_it->second.function_name,
+                  std::move(params_it->second.code),
+                  _executor);
+            } catch (const v8_engine::script_exception& ex) {
+                // add log
+                if (!value->is_deleted) {
+                    // Need to delete runtime from hash_map
+                    _runtimes.erase(topic);
+                    value->cv.broadcast();
+                    co_return nullptr;
+                }
+            }
+
+            // User delete runtime, and we can skip it
+            if (value->is_deleted) {
+                co_return nullptr;
+            }
```

Comment:
can we get rid of `is_deleted` by removing the entry from the runtimes index when its removed?

Reply:
Hm, looks like we can compare pointers to understand what version we have in _runtimes